File: julia/allergies/allergies.jl
const allergens = Dict(
    "eggs"         =>   1,
    "peanuts"      =>   2,
    "shellfish"    =>   4,
    "strawberries" =>   8,
    "tomatoes"     =>  16,
    "chocolate"    =>  32,
    "pollen"       =>  64,
    "cats"         => 128,
)

function allergic_to(score, allergen) 
    haskey(allergens, allergen) && 
    score & allergens[allergen] != 0
end

function allergy_list(score) 
    filter(a -> allergic_to(score, a), keys(allergens))
end

File: clojure/allergies/src/allergies.clj
(ns allergies)

(def ALLERGENS [:eggs :peanuts :shellfish :strawberries
                :tomatoes :chocolate :pollen :cats])

(defn allergies [score]
  (->> (map-indexed list ALLERGENS)     ;;=> '((0 :eggs) (1 :peanuts) ...)
       (filter #(bit-test score (first %)))
       (map second)))

;; Can't call contains? on a LazySeq
(defn lazy-contains? [coll item]
  (boolean (some #(= item %) coll)))

(defn allergic-to? [score allergen]
  (-> (allergies score)
      (lazy-contains? allergen)))


;; Notes:
;; https://clojure-doc.org/articles/language/laziness/
